lib/raft.ts

```diff
@@ -0,0 +1,1381 @@
+    this.saveState(state);
+    this.saveSnapshot(snapshot);
+  }
+
+
+}
+
+export class DummyPersister implements Persister {
+  constructor() { }
+
+  saveState(state: string): void {
+  }
+
+  readSnapshot(): string | undefined {
+    return;
+  }
+
+  readState(): string | undefined {
+    return;
+  }
+
+  saveStateAndSnapshot(state: string, snapshot: string): void {
+
+  }
+
+}
+
+export type State = 'Leader' | 'Follower' | 'Candidate';
+
+const minElectionTimeout = 1000;
+const maxElectionTimeout = 2000;
+
+export interface RaftStateReply {
+  isEnabled: boolean;
+  started: boolean;
+  raftState?: State;
+  leaderBaseURL?: string;
+}
+
+export class Raft {
+  public readonly stateEmitter = new EventEmitter();
+  public readonly isEnabled: boolean;
+  private peers: RaftPeer[]; // RPC end points of all peers
+  private me: number;  // this peer's index into peers[]
+  private state: State = 'Follower';
+  private readonly electionTimeout = Math.floor(Math.random() * (maxElectionTimeout - minElectionTimeout) + minElectionTimeout);
+  private trace = true;
+  private started = false;
+
+  // persistent state
+  private currentTerm: number = 0;
+  private votedFor = -1
+  private log: RaftLogEntry[] = [];
+  private startIndex: number = 0;
+  private startTerm: number = -1;
+
+  // volatile state on all servers
+  private commitIndex: number = -1;
+  private lastApplied: number = -1;
+
+  // volatile state on leaders(Reinitialized after election):
+  private nextIndex: number[] = [];  //  for each server, index of the next log entry to send to that server (initialized to leader last log index + 1)
+  private matchIndex: number[] = []; // for each server, index of highest log entry known to be replicated on server (initialized to 0, increases monotonically)
+  private electionTimeoutId: NodeJS.Timer;
+  private readonly heartbeatInterval: number = Math.round(minElectionTimeout * .75);
+  private heartbeatTimeoutId: NodeJS.Timer;
+  private leaderId: number = -1; // last observed leader id
+  private discardCount: number = 0;
+  private lastSnapshot: Snapshot;
+
+  private persister: Persister;
+  private maxLogSize: number = -1;
+  private syncService: SyncService;
+  private apiml: ApimlConnector;
+
+
+  constructor() {
+    const clusterEnabledEnvVar = process.env.ZOWE_APP_SERVER_CLUSTER_ENABLED;
+    this.isEnabled = clusterEnabledEnvVar === 'TRUE' || clusterEnabledEnvVar === 'YES';
+  }
+
+  async start(apiml: ApimlConnector): Promise<void> {
+    raftLog.info(`Starting peer electionTimeout ${this.electionTimeout} ms heartbeatInterval ${this.heartbeatInterval} ms`);
+    this.apiml = apiml;
+    this.persister = Raft.makePersister();
+    this.maxLogSize = Raft.getMaxLogSize();
+    const { peers, me } = await this.waitUntilZluxClusterIsReady();
+    this.peers = peers;
+    this.me = me;
+    if (me === -1) {
+      raftLog.warn(`unable to find my instance among registered zlux instances`);
+      return;
```

Comment:
It is just an extra check. It must never happen unless the is a bug in APIML. If the instance has registered in Eureka, Eureka must include the instance into the list list of registered instances, otherwise it is a bug.